### PR TITLE
Refactor SDK to use async/await

### PR DESCRIPTION
**Pull request by mmaatttt**
This is unavoidably massive since you can't easily change one part of the code to async without it impacting the rest. In general, anything with a completion handler becomes `async` and returns the type of the completion handler. Any completion handler with a result type also becomes `throws`. Then we `try await` when calling that function and continue to make changes after that. So for example,

```swift
func process(_ activity: Activity, completion: @escaping (Result<QualifyResponse, Error>) -> Void)
// becomes
func process(_ activity: Activity) async throws -> QualifyResponse
```

The public api implications here are pretty minimal and listed in the updated migration guide doc.

A few strategies used:
1. `Task` creates a new unit of asynchronous work. One of the most important decisions that gets made repeatedly is where to introduce a `Task` versus propagating and `async` call.
2. `withCheckedContinuation` wraps a callback interface to convert to an async one.
3. Avoid marking protocols or classes with `@MainActor`. For example, an `AppcuesExperienceAction` may have `@MainActor func execute()`, but it needs to be able to be initialized synchronously from any context. Similarly, the contents of `ModalContextManager` are all `@MainActor` but the class can be inited synchronously by the trait init. Otherwise there's substantially more changes required.
    - The `UIDebugging` and `ScreenCaptureUI` protocol definitions have `@MainActor` requirements because 1) implementations of these will always need it, and 2) it means they can be called synchronously from each other (ie if it were `func showToast(_ toast: DebugToast) async`, then we'd need to `await` calling it from other `@MainActor` funcs).
4. Don't implement actors (yet). There are a few places (`DIContainer`, `ActivityProcessor`, `SdkMetrics`, etc),  where we use `DispatchQueue` to eliminate data races. Actors may be the right solution in a few of these cases, but making that change now would make the scope of this PR drastically larger.
5. In async tests, `waitForExpectations()` isn't supported, and so has to be replaced with `await fulfillment()`. I've taken the opportunity to enforce the order of multiple expectations, which gives us a new level of guarantees on the behaviour being tested. In general, the remove of completion handlers simplifies a lot of the tests.

Some of the nice improvements:
1. The 4 sequential network requests in the screen capturer are way more readable
2. The state machine `transitionAndObserve()` bit is replaced by await-able `transition()` and a `stateStream` for the `dismiss()` case where we may need to wait for the state machine to finish rendering
3. Checking for main thread in the debugger is gone

The most involved changes are with the `ExperienceRenderer` and `ExperienceStateMachine` and there is a bit of refactoring in `♻️ Fix experience dismissal and clean up experience state observing` so it's a standalone commit (marked with a `TODO` in the initial set of changes. The tests pass, and I've tested a bunch, but we'll need to continue hammering on it to ensure we've handled all the edge cases. There's a few places where we were explicitly dispatching to the main thread to get a specific behaviour and I've confirmed those work as expected now without those hacks.

### CODE REVIEW

File: Sources/AppcuesKit/Presentation/Actions/ActionRegistry.swift [hunk 1]
```diff
@@ -9,13 +9,11 @@
...
 
 internal class ActionRegistry {
-    typealias Completion = () -> Void
-
     private var actions: [String: AppcuesExperienceAction.Type] = [:]
 
     private var isProcessing = false
     private var actionQueue: [AppcuesExperienceAction] = [] {
-        didSet { processFirstAction() }
+        didSet { Task { try await processFirstAction() } }
```
Comment (mmaatttt) [inline Sources/AppcuesKit/Presentation/Actions/ActionRegistry.swift:16]: This might be a good candidate to convert to an `AsyncSequence`. Will look into doing so later.

**Review by iujames — COMMENTED** [review]
made it first pass through 73 non-test files so posting any notes thus far, but looking great! lots of nice cleanup

File: Sources/AppcuesKit/Presentation/ExperienceRendering/ExperienceRenderer.swift [hunk 2]
```diff
@@ -113,15 +114,25 @@ internal class ExperienceRenderer: ExperienceRendering, StateMachineOwning {
...
+                errors.append(error)
+            }
         }
 
         // No caching required for modals since they can't be lazy-loaded.
         potentiallyRenderableExperiences.removeValue(forKey: .modal)
+
+        if let error = errors.first {
+            throw error
+        }
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Sources/AppcuesKit/Presentation/ExperienceRendering/ExperienceRenderer.swift:132]: Right. We don't have to throw the error here—it's kinda undefined what would be expected if some presentations succeeds and some fail. You could argue that any success is good enough and there should be no error. We could also do a special "partial success" error that lists what exactly worked and didn't work.

But like you said, it's just logging the error and I think that's a good thing to have.

File: Sources/AppcuesKit/Presentation/Traits/Appcues/AppcuesEmbeddedTrait.swift [hunk 3]
```diff
@@ -44,15 +44,17 @@ internal class AppcuesEmbeddedTrait: AppcuesStepDecoratingTrait, AppcuesContaine
         stepController.padding = NSDirectionalEdgeInsets(paddingFrom: style)
     }
 
+    @MainActor
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Sources/AppcuesKit/Presentation/Traits/Appcues/AppcuesEmbeddedTrait.swift:47]: Correct, they're not required since it's implicit from the protocol definition. I like it being explicit, so I'll probably update all the trait implementations to have the annotations. Especially because these can call helper functions that do need `@MainActor` and it's more obvious if the caller is annotated too.

File: Tests/AppcuesKitTests/Experiences/ExperienceStateMachine+AnalyticsObserverTests.swift [hunk 4]
```diff
@@ -14,59 +14,63 @@ class ExperienceStateMachine_AnalyticsObserverTests: XCTestCase {
...
         appcues = MockAppcues()
         appcues.analyticsPublisher.onPublish = { update in
             self.updates.append(update)
+            self.analyticsExpectation?.fulfill()
         }
         observer = ExperienceStateMachine.AnalyticsObserver(container: appcues.container)
     }
 
-    override func tearDownWithError() throws {
+    override func tearDown() async throws {
```
Comment (iujames) [inline Tests/AppcuesKitTests/Experiences/ExperienceStateMachine+AnalyticsObserverTests.swift:28]: in various spots like this could we update to `override func tearDown()` ? not important, just curious, if we need the throwing cases

and L19 `override func setUp() {`

File: Sources/AppcuesKit/Presentation/Debugger/ScreenCapturing/UIKitElementTargeting.swift [hunk 5]
```diff
@@ -179,16 +179,24 @@ internal extension UIView {
 
         var tabCount = 0
 
-        let children: [AppcuesViewElement] = self.subviews.compactMap { subview -> AppcuesViewElement? in
-            // discard hidden views and subviews within
-            guard !subview.isHidden else { return nil }
+        var children: [AppcuesViewElement] = []
+        for subview in self.subviews {
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Sources/AppcuesKit/Presentation/Debugger/ScreenCapturing/UIKitElementTargeting.swift:183]: Yep. We can roll our own pretty easily if we want though (like https://github.com/JohnSundell/CollectionConcurrencyKit/blob/main/Sources/CollectionConcurrencyKit.swift)

There's an interesting thread on having `asyncForEach` on an `AsyncSequence` (so slightly different context) that makes a good case for just using `for in`: https://forums.swift.org/t/do-we-want-foreach/56929

File: Sources/AppcuesKit/Presentation/CustomComponents/AppcuesExperienceActions.swift [hunk 6]
```diff
@@ -65,7 +65,9 @@ public class AppcuesExperienceActions {
     }
 
     private func enqueue(_ action: AppcuesExperienceAction) {
-        guard let actionRegistry = appcues?.container.resolve(ActionRegistry.self) else { return }
-        actionRegistry.enqueue(actionInstances: [action]) {}
+        Task {
+            guard let actionRegistry = appcues?.container.resolve(ActionRegistry.self) else { return }
+            try await actionRegistry.enqueue(actionInstances: [action])
+        }
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Sources/AppcuesKit/Presentation/CustomComponents/AppcuesExperienceActions.swift:71]: So this is interesting. You absolutely could get actions from multiple user interactions executing out of order (of course the actions for a single user interaction will always execute in order).

`actionRegistry.enqueue` is a misnomer—it's not adding the items to the queue, it's just executing them. We do the same for push actions and pre-step navigation actions where they just execute immediately. At the very least this function should be renamed to not be `enqueue`. But I think there's a good opportunity to refactor further and use an actor and `AsyncStream` instead of janky `isProcessing` flag.

File: Examples/DeveloperCocoapodsExample/CocoapodsExample/DeepLinkNavigator.swift [hunk 7]
```diff
@@ -119,8 +119,12 @@ class DeepLinkNavigator: AppcuesNavigationDelegate {
...
 
     // conforming to AppcuesNavigationDelegate for navigation requests coming from the Appcues SDK
-    func navigate(to url: URL, openExternally: Bool, completion: @escaping (Bool) -> Void) {
-        handle(url: url, openExternally: openExternally, completion: completion)
+    func navigate(to url: URL, openExternally: Bool) async -> Bool {
+        await withCheckedContinuation { continuation in
+            handle(url: url, openExternally: openExternally) { success in
+                continuation.resume(returning: success)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Examples/DeveloperCocoapodsExample/CocoapodsExample/DeepLinkNavigator.swift:127]: Yeah we definitely can rewrite more of the example app, but I like this as an example for folks updating their SDK, but not necessarily looking to do more refactoring.

File: Tests/AppcuesKitTests/Actions/AppcuesCloseActionTests.swift [hunk 8]
```diff
@@ -28,54 +28,46 @@ class AppcuesCloseActionTests: XCTestCase {
...
-        action?.execute(completion: { completionCount += 1 })
+        try await action?.execute()
 
         // Assert
-        XCTAssertEqual(completionCount, 1)
         XCTAssertEqual(dismissCount, 1)
     }
 
-    func testExecuteCompletesWithoutAppcuesInstance() throws {
+    func testExecuteCompletesWithoutAppcuesInstance() async throws {
```
[the remark this answers is not in the document]
Comment (mmaatttt) [reply Tests/AppcuesKitTests/Actions/AppcuesCloseActionTests.swift:63]: Will do `testExecuteThrowsWithoutAppcuesInstance()` 👍

File: Tests/AppcuesKitTests/Experiences/ExperienceStateMachine+AnalyticsObserverTests.swift (hunk 4 above)
Comment (mmaatttt) [reply Tests/AppcuesKitTests/Experiences/ExperienceStateMachine+AnalyticsObserverTests.swift:28]: It's a bit goofy how XCTest has evolved over the years. You can do any of
```swift
func setUp()
```

File: Sources/AppcuesKit/Presentation/ExperienceRendering/ExperienceRenderer.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (iujames) [reply Sources/AppcuesKit/Presentation/ExperienceRendering/ExperienceRenderer.swift:132]: yep, agreed!
